### ISSUE TITLE
[BRGEMM Ukernel] Call of execute with empty post op parameter

### ISSUE THREAD

**Issue by Devjiu** [oneapi-src/oneDNN#2179]
# Summary
During modification of initial example for brgemm ukernel usage I removed post-ops from configuration but kept `execution` call with default value of post-ops parameters. In this configuration accumulation buffer is ignored and don't passed into output buffer. 
In general, issue is behavior of this [execute](https://github.com/oneapi-src/oneDNN/blob/04ce01fe5461f45033daa78cebdbb6646da9a808/include/oneapi/dnnl/dnnl_ukernel.hpp#L345)  with default attributes don't match with behavior of raw brgemm kernel's [execute](https://github.com/oneapi-src/oneDNN/blob/04ce01fe5461f45033daa78cebdbb6646da9a808/include/oneapi/dnnl/dnnl_ukernel.hpp#L321).

# Version
onednn_verbose,v1,info,oneDNN v3.7.0 (commit 2f8619b0b9bb64e75a8da8f49f70ae226f841bbc)
onednn_verbose,v1,info,cpu,runtime:OpenMP,nthr:128
onednn_verbose,v1,info,cpu,isa:Intel AVX-512 with Intel DL Boost
onednn_verbose,v1,info,gpu,runtime:none
onednn_verbose,v1,info,graph,backend,0:dnnl_backend

# Environment
oneDNN includes hardware-specific optimizations and may behave
differently on depending on the compiler and build environment. Include
the following information to help reproduce the issue:
* CPU make and model 
```
 Architecture:           x86_64
  CPU op-mode(s):       32-bit, 64-bit
  Address sizes:        52 bits physical, 57 bits virtual
  Byte Order:           Little Endian
CPU(s):                 128
  On-line CPU(s) list:  0-127
Vendor ID:              GenuineIntel
  Model name:           Intel(R) Xeon(R) Platinum 8352Y CPU @ 2.20GHz
    CPU family:         6
    Model:              106
    Thread(s) per core: 2
    Core(s) per socket: 32
    Socket(s):          2
    Stepping:           6
    CPU max MHz:        3400.0000
    CPU min MHz:        800.0000
    BogoMIPS:           4400.00
    Flags:              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts
                        rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_t
                        imer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid c
                        qm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_l
                        ocal split_lock_detect wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx
                        512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Caches (sum of all):
  L1d:                  3 MiB (64 instances)
  L1i:                  2 MiB (64 instances)
  L2:                   80 MiB (64 instances)
  L3:                   96 MiB (2 instances)
NUMA:
  NUMA node(s):         2
  NUMA node0 CPU(s):    0-31,64-95
  NUMA node1 CPU(s):    32-63,96-127
Vulnerabilities:
  Gather data sampling: Mitigation; Microcode
  Itlb multihit:        Not affected
  L1tf:                 Not affected
  Mds:                  Not affected
  Meltdown:             Not affected
  Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
  Retbleed:             Not affected
  Spec rstack overflow: Not affected
  Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
  Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
  Srbds:                Not affected
  Tsx async abort:      Not affected
```
* OS version 
```
Linux jupyter-dmitriim 5.15.0-97-generic #107-Ubuntu SMP Wed Feb 7 13:26:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```
* Compiler version 
```
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
```
* CMake version
```
cmake version 3.30.1
```
* CMake output log
```
$ cmake .. -DDNNL_EXPERIMENTAL_UKERNEL=ON
CMake Deprecation Warning at CMakeLists.txt:17 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- CMAKE_BUILD_TYPE is unset, defaulting to Release
-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- DNNL_TARGET_ARCH: X64
-- DNNL_LIBRARY_NAME: dnnl
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Could NOT find Doxyrest (missing: DOXYREST_EXECUTABLE)
CMake Warning (dev) at cmake/Sphinx.cmake:25 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  cmake/doc.cmake:28 (include)
  CMakeLists.txt:135 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found PythonInterp: .venv/bin/python (found suitable version "3.10.12", minimum required is "2.7")
-- Could NOT find Sphinx (missing: SPHINX_EXECUTABLE)
-- Found Git: /usr/bin/git (found version "2.34.1")
-- Enabled testing coverage: CI
-- Enabled workload: TRAINING
-- Enabled primitives: ALL
-- Enabled primitive CPU ISA: ALL
-- Enabled primitive GPU ISA: ALL
-- Enabled GeMM kernels ISA: ALL
-- Primitive cache is enabled
-- Experimental functionality for ukernels is enabled
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Graph component is enabled
-- Graph compiler backend is disabled.
-- Configuring done (0.8s)
-- Generating done (0.4s)
-- Build files have been written to: oneDNN/build
```
* git hash 
```
based on 2f8619b0b9bb64e75a8da8f49f70ae226f841bbc 
```

# Steps to reproduce
PR to reproduce: https://github.com/Devjiu/oneDNN/pull/1 . It modifies [examples/ukernels/cpu_brgemm.cpp](https://github.com/Devjiu/oneDNN/pull/1/files#diff-212195ddf2d10e7c53b5b26281b7c3bfe233bdd82c9efade5c8f51cb9395ceb3)
Just run this test with comparison with reference output.   With execute, used [here](https://github.com/Devjiu/oneDNN/pull/1/files#diff-212195ddf2d10e7c53b5b26281b7c3bfe233bdd82c9efade5c8f51cb9395ceb3L318) output result from D_mem don't have data from accumulator buffer C_ptr. 

# Observed behavior
Dump with reduced size and enabled verbose in this [comment](https://github.com/Devjiu/oneDNN/pull/1#issuecomment-2429019344). 

# Expected behavior
I expect that those calls: 
```
brg_po.execute(A_ptr, B_base_ptr, A_B_po_offsets, C_ptr, scratchpad.data());
```
and 
```
brg_po.execute(A_ptr, B_base_ptr, A_B_po_offsets, C_ptr,
                 D_mem.get_data_handle(), scratchpad.data());
```

will have the same result. (`C_ptr` in first case, `D_mem` in second one)

**Comment by mgouicem** [comment]
Hey @Devjiu thanks for the report. Indeed, when calling the extended execute function, without post-op nor conversion, the C matrix is not read (it seems that internally, it reads D and accumulate in it in that situation).

You can workaround the issue by using the [minimal version of execute ](https://github.com/oneapi-src/oneDNN/blob/04ce01fe5461f45033daa78cebdbb6646da9a808/include/oneapi/dnnl/dnnl_ukernel.hpp#L321) (without D matrix) when no post-op nor data conversion is needed.

On oneDNN side, we will clarify this behavior in documentation. We can likely add internal checks to avoid this issue for future users.

**Comment by dzarukin** [comment]
Hi @Devjiu, we've added a check inside extended execute call to verify if it is legit to make this call with correspondent ukernel object or not. There's also a new API call that can be used to identify if execute is legit ahead of calling it. I've updated the example accordingly with it, and with the patch shared in this report it's now passing if `brg_po` object drops post-ops.
Feel free to take a look and share your feedback if any. Thanks.
